src/symbolic/simplify_program.jl

```diff
@@ -111,24 +111,55 @@ function get_program_rules(alg, shash)
+                Postwalk(@rule assign(access(~lhs, ~m, ~j...), ~f::iscollapsible(alg), ~rhs) => begin  
+                             if !(lhs in decl_in_scope)
+                                 collapsed(alg, idx, ext.val, access(lhs, m, j...), f, rhs)
+                             end
+                         end)(body)
             end
         end),
-        (@rule loop(~i, ~a::isvirtual, assign(access(~b, updater(), ~j...), +, ~d)) => begin
-            if i ∉ j && i ∉ getunbound(d)
-                assign(access(b, updater(), j...), +, call(*, measure(a.val), d))
+
+        # Lifting sieve (TODO multiple sieves in the block) 
+        (@rule loop(~idx, ~ext::isvirtual, sieve(~cond, ~body)) => begin
+            if idx ∉ getunbound(cond)
+                sieve(cond, loop(idx, ext, body))
+            end
+        end),
+
+        # Bottom-up reduction1
+        (@rule loop(~idx, ~ext::isvirtual, assign(access(~lhs, ~m, ~j...), ~f::iscollapsible(alg), ~rhs)) => begin
+            if idx ∉ j && idx ∉ getunbound(rhs)
+                decl_in_scope = filter(!isnothing, map(node-> if @capture(node, declare(~tns, ~init)) tns end, PostOrderDFS(rhs)))
+                if !(lhs in decl_in_scope)
+                    collapsed(alg, idx, ext.val, access(lhs, m, j...), f, rhs)
+                end
             end
         end),
 
+        # Bottom-up reduction2
+        (@rule loop(~idx, ~ext::isvirtual, block(~s1..., assign(access(~lhs, ~m, ~j...), ~f::iscollapsible(alg), ~rhs), ~s2...)) => begin 
+           if ortho(getroot(lhs), s1) && ortho(getroot(lhs), s2)
+               if idx ∉ j && idx ∉ getunbound(rhs)
+                   decl_in_scope = filter(!isnothing, map(node-> if @capture(node, declare(~tns, ~init)) tns end, PostOrderDFS(rhs)))
+                   if !(lhs in decl_in_scope)
+                       collapsed_body = collapsed(alg, idx, ext.val, access(lhs, m, j...), f, rhs)
+                       block(collapsed_body, loop(i, ext, sequence(s1..., s2...)))
+                   end
+               end
+           end
+        end),
+        
         (@rule assign(~a, ~op, cached(~b, ~c)) => assign(a, op, b)),
 
-        (@rule loop(~i, ~ext::isvirtual, assign(access(~a, ~m), $(literal(+)), ~b::isliteral)) =>
-            assign(access(a, m), +, call(*, b, measure(ext.val)))
-        ),
-        (@rule loop(~i, ~ext::isvirtual, block(~s1..., assign(access(~a, ~m), $(literal(+)), ~b::isliteral), ~s2...)) => if ortho(getroot(a), s1) && ortho(getroot(a), s2)
-            block(assign(access(a, m), +, call(*, b, measure(ext.val))), loop(i, ext, block(s1..., s2...)))
+        (@rule loop(~i, ~a::isvirtual, assign(access(~b, updater(), ~j...), ~f::isidempotent(alg), ~c)) => begin
+            if i ∉ j && i ∉ getunbound(c)
+                sieve(call(>, measure(a.val), 0), assign(access(b, updater(), j...), f, c))
```

Comment:
`sieve(call(>, measure(a.val), 0), ...)` could be moved to `collapsed` as well

Reply:
What is the purpose of using a `sieve` on idempotent functions? It seems unnecessary to use a `sieve` on **non-idempotent** functions. I understand that removing the `sieve` could result in several failed tests.

Are there any differences in lowering between idempotent and non-idempotent functions?

Reply:
we use sieve to avoid the empty interval.

```
max(a) = a
max(a, b) = max(a, b)
max(a, b, b) = max(a, b)
max(a, b, b, b) = max(a, b)
```

however, when we collapse +, the * operator handles the empty interval just fine:

```
a = a + 0 * b
a + b = a + 1 * b
a + b + b = a + 2 * b
```